File: C-Plus-Plus/bits_manipulation/Hamming_Distance.cpp
// Hamming Distance
/*
For given x and y, return the number of positions where their values differ in their binary representations as a 32-bit integer.

Input
x = 10
y = 6
Output
2
*/

/* Implementation: x = 10 (1010) and y = 6 (0110) if we apply xor to x and y the resultant ans = 12 (1100).
XOR is 0 when 2 bits are equal and 1 when 2 bits contrasts
Number of bits in ans = 2
*/

#include<bits/stdc++.h> 
using namespace std;

int solve(int x, int y) {
    int ans = x^y;
    return __builtin_popcount(ans); // O(N) time complexity
}

int main(){
    int x,y; cin>>x>>y;
    cout<<solve(x,y)<<endl;
}

File: Python/math/Hamming_distance.py
"""
Problem Statement : To find Hamming Distance between two numbers.
                    Hamming distance? The hamming distance is the value differ in two binary 
                    representations as a 32-bit integer.

Intution: If num1 = 7(0111) and num2 = 8(1000), the bit differance between 
          these two numbers will be 4.

"""

def hamming_distance(num1, num2):

    #To store value of the differnce
    bitdiff = 0
    
    for i in range(31, -1, -1):
        
        #right shift for both the numbers for i AND 1 times
        binary_num1 = num1>>i&1
        binary_num2 = num2>>i&1

        #checking if binary of both are not equal
        if binary_num1 != binary_num2:
            bitdiff += 1
        #i it is equal bitdiff will be same and loop will continue

    return bitdiff

def main():

    #User input for a both numbers
    num1 = int(input("Enter 1st number: "))
    num2 = int(input("Enter 2nd number: "))

    #function call for getting hamming distance for the given two numbers.
    print("Hamming distance between these two numbers is: ", hamming_distance(num1, num2))

main()

'''
Time complexity :O(n)
Space complexity :O(1)
'''

"""
Sample Input / Output:

Enter 1st number: 7
Enter 2nd number: 8
Hamming distance between these two numbers is:  4

Enter 1st number: 1
Enter 2nd number: 6
Hamming distance between these two numbers is:  3

"""
